### PR TITLE
feat(lsp): configurable diagnostic severity

### PR DESCRIPTION
**Pull request by matoous**
Allow severity of diagnostic messages to be configured.
E.g. allow turning of Hint level diagnostics.

Fixes: https://github.com/helix-editor/helix/issues/1007

### CODE REVIEW

File: helix-view/src/document.rs [hunk 1]
```diff
@@ -88,7 +88,7 @@ pub struct Document {
 
     syntax: Option<Syntax>,
     // /// Corresponding language scope name. Usually `source.<lang>`.
-    pub(crate) language: Option<Arc<LanguageConfiguration>>,
+    pub language: Option<Arc<LanguageConfiguration>>,
```
Comment (matoous) [inline helix-view/src/document.rs:91]: Not sure if this is a good idea?

Comment (archseer) [reply helix-view/src/document.rs:91]: Use `doc.language()`

Comment (matoous) [reply helix-view/src/document.rs]: Thanks, totally missed that there is `language_config()`. Done.

File: helix-core/src/syntax.rs [hunk 2]
```diff
@@ -63,6 +64,8 @@ pub struct LanguageConfiguration {
 
     #[serde(default)]
     pub auto_format: bool,
+    #[serde(default)]
+    pub diagnostic_severity: Severity,
```
Comment (archseer) [inline helix-core/src/syntax.rs:68]: This also needs a description in `book/configuration.md`

Comment (matoous) [reply helix-core/src/syntax.rs:68]: Added.

File: book/src/guides/adding_languages.md [hunk 3]
```diff
@@ -27,18 +27,19 @@ directory](../configuration.md).
...
+| name                | The name of the language                                      |
+| scope               | A string like `source.js` that identifies the language. Currently, we strive to match the scope names used by popular TextMate grammars and by the Linguist library. Usually `source.<name>` or `text.<name>` in case of markup languages |
+| injection-regex     | regex pattern that will be tested against a language name in order to determine whether this language should be used for a potential [language injection][treesitter-language-injection] site. |
+| file-types          | The filetypes of the language, for example `["yml", "yaml"]`  |
+| shebangs            | The interpreters from the shebang line, for example `["sh", "bash"]` |
+| roots               | A set of marker files to look for when trying to find the workspace root. For example `Cargo.lock`, `yarn.lock` |
+| auto-format         | Whether to autoformat this language when saving               |
+| diagnostic_severity | Minimal severity of diagnostic for it to be displayed. (Allowed values: `Error`, `Warning`, `Info`, `Hint`) |
```
Comment (matoous) [inline book/src/guides/adding_languages.md:39]: This is the new entry, the rest is just formatting

**Review by dead10ck — APPROVED** [review]
Thanks!

File: helix-term/src/application.rs [hunk 4]
```diff
@@ -415,19 +416,28 @@ impl Application {
                                         return None;
                                     };
 
+                                    let severity =
+                                        diagnostic.severity.map(|severity| match severity {
+                                            DiagnosticSeverity::ERROR => Error,
+                                            DiagnosticSeverity::WARNING => Warning,
+                                            DiagnosticSeverity::INFORMATION => Info,
+                                            DiagnosticSeverity::HINT => Hint,
+                                            severity => unimplemented!("{:?}", severity),
```
[the remark this answers is not in the document]
Comment (matoous) [reply helix-term/src/application.rs:425]: I think this is not going to change (at least in foreseeable future). I changed the statement to `unreachable!`, let me know if I should revert it.

File: book/src/guides/adding_languages.md (hunk 3 above)
Comment (archseer) [reply book/src/guides/adding_languages.md:39]: `diagnostic-severity`, we remap the name to kebab-case via serde

**Review by sudormrfbin — APPROVED** [review]
LGTM !

File: book/src/guides/adding_languages.md (hunk 3 above)
Comment (matoous) [reply book/src/guides/adding_languages.md]: sorry, forgot, fixed now

**Comment by sudormrfbin** [comment]
@matoous seems like an implementation of `shrink_selection` crept through in the latest commit :P

**Comment by matoous** [comment]
@sudormrfbin yes, sorry 🙈 fixed now

**Comment by archseer** [comment]
Thanks! Looks good now 👍

**Pull request merged**